aip/general/0180.md

```diff
@@ -139,6 +139,30 @@ code, as such changes will be seen as breaking by those users.
 whether a proposed change is likely to break users, and an expansive reading of
 this guidance could ostensibly prevent _any_ change (which is not the intent).
 
+#### Field behavior compatibility
+
+The behavior of field within an API, whether it is required, whether it is only
+presented as output and not accepted as input, or vice versa, is documented with
+the `google.api.field_behavior` annotation as described in [AIP-203][]. Changing
+the value of this annotation can represent a semantic change in the API that is
+percevied as incompatible for existing clients. The following are examples of
+backwards incompatible changes with `google.api.field_behavior`:
+
+* Adding `REQUIRED` to an existing field previously considered `OPTIONAL`
+(implicitly or otherwise).
+* Adding `OUTPUT_ONLY` to an existing field previously accepted as input
+* Removing `OUTPUT_ONLY` from an existing field previously ignored as input
```

Comment:
what makes this backwards-incompatible? I think it's not going to be a blocker, but it seems like a parameter could become writable.

Reply:
This is a sneaky one imo. The guidance for `OUTPUT_ONLY` is as follows:

> The use of `OUTPUT_ONLY` indicates that the field is provided in responses, but
that including the field in a message in a request does nothing (the server
**must** clear out any value in this field and **must not** throw an error as a
result of the presence of a value in this field on input). Similarly, services
**must** ignore the presence of output only fields in update field masks.

If I was a user naively setting an `OUTPUT_ONLY` field and not getting any errors so I'm not changing my code, I would be surprised if suddenly that field was being accepted as input and either broke my previously working calls or broke my resource's behavior in some way.

IDK how it would work in terraform, but if I could, in my terraform config, set an output only field, and terraform ignores it in the response during reconciliation b.c its output_only, but suddenly it starts being accepted as input by the server, what happens? Seems sticky.

This "silently ignore output_only fields in the request" is import for the read-modify-write flow, but could be dangerous for compatibility. `IMMUTABLE` doesn't have this problem because our guidance states that requesting to update an immutable field should produce an error.

Reply:
Terraform won't let you specify output-only fields in your config, but I think the path from output-only to settable is scary in general.

For example, a user may be using the value of an output-only field as an input to another resource. Then the field is marked as settable, so Terraform may (depending on some specifics of the field's behavior) try to remove the existing value for the field because it isn't present in the user's config, breaking the downstream resource

It would definitely require some thought as to what is acceptable

Reply:
> For example, a user may be using the value of an output-only field as an input to another resource. Then the field is marked as settable, so Terraform may (depending on some specifics of the field's behavior) try to remove the existing value for the field because it isn't present in the user's config, breaking the downstream resource.

This seems like a good use case for a field mask to help decide what is cleared vs not, don't you think? 

I think that workflow is definitely dangerous, and I don't think this happens enough that it's super low friction to introduce a new input value the `OUTPUT_ONLY` parameter becomes configurable.

Okay, it's fine to leave this for now, and let's see what user feedback brings.